scoring/src/main/scala/ai/h2o/sparkling/ml/models/H2OMOJOBaseCache.scala

```diff
@@ -0,0 +1,79 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package ai.h2o.sparkling.ml.models
+
+import org.apache.spark.expose.Logging
+import org.apache.spark.sql.SparkSession
+
+import scala.collection.mutable
+
+trait H2OMOJOBaseCache[B, M] extends Logging {
+  private object Lock
+
+  private val pipelineCache = mutable.Map.empty[String, B]
```

Comment:
WeakHashMap?

Reply:
I really like simplicity of your proposal. But if I'm not mistaken, `WeakHashMap` evicts an entry by a GC thread if there are no strong reference to its key. IMHO, this situation could happen in the preparation period between execution of two tasks, especially when `--executor-cores=1`.

Or am I missing something in your proposal?

Reply:
right, good point! weakhashmap would optimize evection rate, but we need optimize hit rate (expectation is that multiple tasks will use mojo).